### ISSUE TITLE
AllowEmptyStrings not generated in CSharp

### ISSUE THREAD

**Issue by cortex93** [RSuter/NSwag#1438]
Hi,

NJsonSchema has support for DataAnnotations Required[AllowEmptyStrings=true] but for some reason, when the DTO classes in generated from the NSwag generator, it does not work. Using the NJsonSchema generator directly works.

I saw that NJsonSchema has two unit test and one of them When_property_is_required_then_required_attribute_is_rendered does not produce AllowEmptyStrings. But I can't see the difference with When_property_is_required_then_required_attribute_is_rendered_in_Swagger_mode.

Is there a settings for that in NSwag ?

**Comment by RSuter** [comment]
NJS has a setting SchemaType (JsonSchema, Swagger2, OpenApi3). The problem here is that json schema has the concept of null and required wheras swagger only knows required... this is why we cant handle this scenario correctly in swagger - maybe with x-nullable custom extension?

**Comment by RSuter** [comment]
Not sure if the code generators actually generate AllowEmptyStrings

**Comment by RSuter** [comment]
![image](https://user-images.githubusercontent.com/2603405/42319800-54fa6a42-8053-11e8-9260-ee3449eb5a05.png)

**Comment by RSuter** [comment]
Just checked this AllowEmptyStrings = false will set MinLength to 1... and it's only doing this for JsonSchema (why? Probably because it would set it almost always as the default of AllowEmptyStrings is false)

![image](https://user-images.githubusercontent.com/2603405/42319917-ca2d087e-8053-11e8-896d-8eff27c7884a.png)

So maybe we should set this for all schema types and in the code generator we render AllowEmptyStrings = true/false based on the MinLength and when the schema type is string?

**Comment by RSuter** [comment]
Ref: https://github.com/RSuter/NJsonSchema/issues/617

Actually we only need to remove the schema type check

**Comment by cortex93** [comment]
Just to be in phase, I generate the c# dto.
In NJsonSchema, the AllowEmptyStrings property on the PropertyModel seems good to me and it make no difference on the schema type.
![image](https://user-images.githubusercontent.com/4573198/42322906-7ff703e4-805e-11e8-96ac-c896ce727595.png)

**Comment by RSuter** [comment]
Huh?
Current implementation is with a schema type check: 

![image](https://user-images.githubusercontent.com/2603405/42322993-c4f355a6-805e-11e8-8a0f-461ac2559575.png)

**Comment by RSuter** [comment]
Not sure if this is a good idea because external specs (not generated by nswag) will now always render AllowEmptyStrings = true which is probably not intended... maybe we need a setting for this

**Comment by cortex93** [comment]
In swagger empty string "" is a valid string unless minLength or pattern is specified.
So it basicly the same interpretation of empty string as JsonSchema.

nullable appears in swagger 3, but it should not affect the AllowEmptyString behavior.

**Comment by RSuter** [comment]
Ok, then the current implementation is correct even if most people won't expect it like that? I bet most will not set MinLength to 1 even if they want to avoid empty strings :)

**Comment by cortex93** [comment]
Though, it is a breaking changes for sure as MVC will now accept value people is not used to.

**Comment by RSuter** [comment]
Yeah, but 90% of the time the generated code is used in the client and these attributes are not used anyway... the server will still behave as before.

**Comment by cortex93** [comment]
Ok, we are all set then.
Thanks

**Comment by RSuter** [comment]
I hate making these breaking changes but it's always a compromise between a breaking change and an additional setting (more complexity) which needs to be maintained forever :)

**Comment by paulomorgado** [comment]
I'm not knowledgeable of this subject enough to have an educated opinion. I am, however, very sensible to breaking changes. What are we talking about here?

**Comment by RSuter** [comment]
This will change how [Required] attributes are generated. 

**Previously:**

Schema generator: `[Required(AllowEmptyStrings=false)` and `[Required(AllowEmptyStrings=true)` generated `required` but no `minlength=1`

Code generator: 

- `required` but no `minlength=1` => `[Required]`
- `required` AND `minlength=1` => `[Required][MinLenght(1)]`

**Now:**

Schema generator: 

- `[Required(AllowEmptyStrings=false)` and `` and `[Required(AllowEmptyStrings=true)` generated `required` but no `minlength=1`

Code generator: 

- `required` but no `minlength=1` => `[Required(AllowEmptyStrings = true)]`
- `required` AND `minlength=1` => `[Required(AllowEmptyStrings = false)]`

This might now be a problem for external schemas where minlength is not 1 => will generate `[Required(AllowEmptyStrings = true)]` which is correct but maybe not expected.

/cc @paulomorgado

**Comment by andrewmorris-me** [comment]
I'm a bit confused as to what is being generated. For example:

```
public class Idea
    {
        [Required]
        [StringLength(10)]
        public string Name { get; set; }

        [Required]
        [MaxLength(5000)]
        public string Description { get; set; }
    }
```

Will generate the following:

```
[Newtonsoft.Json.JsonProperty("name", Required = Newtonsoft.Json.Required.Always)]
[System.ComponentModel.DataAnnotations.Required(AllowEmptyStrings = true)]
[System.ComponentModel.DataAnnotations.StringLength(10)]
public string Name { get; set; }
    
[Newtonsoft.Json.JsonProperty("description", Required = Newtonsoft.Json.Required.Always)]
[System.ComponentModel.DataAnnotations.Required]
[System.ComponentModel.DataAnnotations.StringLength(5000, MinimumLength = 1)]
```

Effectively, are they not the same thing? Why would one be required with an empty string, but the other is not?
